modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -3114,102 +3130,88 @@ void GDScriptAnalyzer::reduce_identifier_from_base(GDScriptParser::IdentifierNod
 	}
 
 	GDScriptParser::ClassNode *base_class = base.class_type;
+	List<GDScriptParser::ClassNode *> script_classes;
+	bool is_base_instance = !base.is_meta_type;
 
-	// TODO: Switch current class/function/suite here to avoid misrepresenting identifiers (in recursive reduce calls).
-	while (base_class != nullptr) {
-		if (base_class->identifier && base_class->identifier->name == name) {
-			p_identifier->set_datatype(base_class->get_datatype());
+	if (base_class != nullptr) {
+		get_class_node_current_scope_classes(base_class, &script_classes);
+	}
+
+	for (GDScriptParser::ClassNode *script_class : script_classes) {
+		if (p_base == nullptr && script_class->identifier && script_class->identifier->name == name) {
+			reduce_identifier_from_base_set_class(p_identifier, script_class->get_datatype(), base.script_path);
 			return;
 		}
 
-		if (base_class->has_member(name)) {
-			resolve_class_member(base_class, name, p_identifier);
+		if (script_class->has_member(name)) {
+			resolve_class_member(script_class, name, p_identifier);
 
-			GDScriptParser::ClassNode::Member member = base_class->get_member(name);
-			p_identifier->set_datatype(member.get_datatype());
+			GDScriptParser::ClassNode::Member member = script_class->get_member(name);
 			switch (member.type) {
-				case GDScriptParser::ClassNode::Member::CONSTANT:
+				case GDScriptParser::ClassNode::Member::CONSTANT: {
+					p_identifier->set_datatype(member.get_datatype());
 					p_identifier->is_constant = true;
 					p_identifier->reduced_value = member.constant->initializer->reduced_value;
 					p_identifier->source = GDScriptParser::IdentifierNode::MEMBER_CONSTANT;
 					p_identifier->constant_source = member.constant;
-					break;
-				case GDScriptParser::ClassNode::Member::ENUM_VALUE:
+					return;
+				}
+
+				case GDScriptParser::ClassNode::Member::ENUM_VALUE: {
+					p_identifier->set_datatype(member.get_datatype());
 					p_identifier->is_constant = true;
 					p_identifier->reduced_value = member.enum_value.value;
 					p_identifier->source = GDScriptParser::IdentifierNode::MEMBER_CONSTANT;
-					break;
-				case GDScriptParser::ClassNode::Member::ENUM:
-					if (p_base != nullptr && p_base->is_constant) {
+					return;
+				}
+
+				case GDScriptParser::ClassNode::Member::ENUM: {
+					p_identifier->set_datatype(member.get_datatype());
+					if (base.is_constant) {
 						p_identifier->is_constant = true;
 						p_identifier->source = GDScriptParser::IdentifierNode::MEMBER_CONSTANT;
```

Comment:
@vonagam I rebased the PR and used `base.is_constant` instead of `p_base->is_constant`.

Reply:
Oh. Two notes.

a) I would have preferred `base.is_meta_type` as more self-explanatory than `base.is_constant`.
But since there is no possibility for a constant instance at the moment it does not matter much.

b) After seeing this check though, it seems wrong to have. It is the same deal that we had with a nested classes. Since all static things are exposed through an instance - constants, enum values, nested classes, static functions - it is strange not to expose an enum too. And the fact that there is a call to `set_datatype` before `if` means that it is exposed. So I checked:
```gdscript
class A:
  enum Hello { World }

func _ready() -> void:
  var a := A.new()
  print(a.Hello) # { "World": 0 }
  print(A.Hello) # <null>
```
I think `if` guard should be removed and `reduced_value` needs to be set.